### PR TITLE
Kernel: Allow switching to IOAPIC mode even without enabling SMP

### PR DESCRIPTION
**Pull request by supercomputer7**
This small change allows to use the IOAPIC by default without to enable
SMP mode, which emulates Uni-Processor setup with IOAPIC instead of
using the PIC.

This opens the opportunity to utilize other types of interrupts like MSI
and MSI-X interrupts.

### CODE REVIEW

**Comment by aajnfl** [comment]
Why not just remove PIC support altogether? Is Serenity supposed to be able to work on pre-486 machines?

**Comment by supercomputer7** [comment]
> 

Interesting Question.
My first thought was to dismiss it but to be honest, it's actually a valid question. We don't drop support (for now) of the (Intel 8259) PIC because:
1. On bare metal IOAPIC support can be fragile, especially on platforms with bogus multiprocessor tables. Until we have proper support for ACPI AML so we can figure IRQ routing with relying on the MP table, we can't really drop support for the PIC.
Please note that on QEMU the MP table is mostly OK but you don't get to understand new routes of new installed PCIe devices, which ACPI AML solves.
2. On some old machines, (#1500), IRQ0 is not connected to the IOAPIC which means we still need the PIC. However, I never encountered such motherboard so it's not a strong reason to keep the PIC. The first point with this justifies it for now.

So, yes, we could remove PIC support in the future, but not yet.

**Comment by supercomputer7** [comment]
Apparently the CI found a bug in the `InterruptManagement` code. Time to fix it then :)

**Comment by supercomputer7** [comment]
Another commit was added, let's see if CI is green now :)

**Comment by dascandy** [comment]
I have a small kernel project that does not support PIC or ACPI, and it runs fine with just MSI/MSI-X using IOAPIC and local APIC.

File: Kernel/Interrupts/InterruptManagement.cpp [hunk 1]
```diff
@@ -206,16 +207,20 @@ UNMAP_AFTER_INIT void InterruptManagement::locate_apic_data()
...
-                interrupt_override_entry->flags);
+                global_system_interrupt,
+                flags);
 
             dbgln("Interrupts: Overriding INT {:#x} with GSI {}, for bus {:#x}",
                 interrupt_override_entry->source,
-                interrupt_override_entry->global_system_interrupt,
-                interrupt_override_entry->bus);
+                global_system_interrupt,
+                flags);
```
Comment (dascandy) [inline Kernel/Interrupts/InterruptManagement.cpp:223]: Was this intentional (replacing bus with flags) ?

File: Kernel/Interrupts/APIC.cpp [hunk 2]
```diff
@@ -270,38 +270,39 @@ UNMAP_AFTER_INIT bool APIC::init_bsp()
...
-            if ((plapic_entry->flags & 0x1) != 0)
-                m_processor_enabled_cnt++;
-        } else if (madt_entry->type == (u8)ACPI::Structures::MADTEntryType::Local_x2APIC) {
-            // Only used for APID IDs >= 255
-            auto* plx2apic_entry = (const ACPI::Structures::MADTEntries::ProcessorLocalX2APIC*)madt_entry;
-            dbgln_if(APIC_DEBUG, "APIC: AP found @ MADT entry {}, processor ID: {}, x2APIC ID: {}, flags: {:#08x}", entry_index, plx2apic_entry->acpi_processor_id, plx2apic_entry->apic_id, plx2apic_entry->flags);
-            m_processor_cnt++;
-            if ((plx2apic_entry->flags & 0x1) != 0)
-                m_processor_enabled_cnt++;
+    if (kernel_command_line().is_smp_enabled()) {
```
Comment (dascandy) [inline Kernel/Interrupts/APIC.cpp:273]: Not doing MADT parsing for IOAPIC or redir entries here? This smells wrong to me, and I'm not seeing where else you're finding the IOAPICs.

File: Kernel/Interrupts/InterruptManagement.cpp (hunk 1 above)
Comment (supercomputer7) [reply Kernel/Interrupts/InterruptManagement.cpp:223]: Oops, that is a mistake. Thanks for putting a note on this!

File: Kernel/Interrupts/APIC.cpp (hunk 2 above)
Comment (supercomputer7) [reply Kernel/Interrupts/APIC.cpp:273]: I find IOAPICs here:
https://github.com/SerenityOS/serenity/blob/f151a0d72057f9ce2be459e4c4fab9093f349ffe/Kernel/Interrupts/InterruptManagement.cpp#L200-L206

and I find redirections here:
https://github.com/SerenityOS/serenity/blob/f151a0d72057f9ce2be459e4c4fab9093f349ffe/Kernel/Interrupts/InterruptManagement.cpp#L207-L219

So we don't need to do that in the `APIC.cpp` file.

**Comment by supercomputer7** [comment]
> I have a small kernel project that does not support PIC or ACPI, and it runs fine with just MSI/MSI-X using IOAPIC and local APIC.

It's definitely possible to do that with MP tables and trying to guess the right IRQ line. We don't do that and we want ACPI support. When we have ACPI AML support I think we can remove the i8259 code for good.

**Comment by supercomputer7** [comment]
We should also merge this because it will give a motivation (hopefully also for me) to create AML parser just to be able to decode the IRQ routing from DSDT blob. I feel like we are being stuck on this topic for way too long, and we should just put something barebones to work first and that's it as for a start.

**Comment by supercomputer7** [comment]
This fixes #10951.

**Pull request merged**